### PR TITLE
Improve BenchmarkParquetReader

### PR DESCRIPTION
**Pull request by yiweiHeOSS**
This PR is to fix https://github.com/facebookincubator/velox/issues/6247.

So far the Benchmark test has 4 types, this PR added more types to the BenchmarkParquetReader test:

1. ShortDecimalType
2. LongDecimalType
3. VARCHAR

We should test them with different filter rates and null rates just like the previous tests in BenchmarkParquetReader.

Also, I noticed we have never done the filter test for the type HUGEINT (int128_t) before, which is the actual type of LongDecimalType. So this PR also implemented the code to generate the filter of the type HUGEINT (int128_t) and modify the code to generate the data of the type HUGEINT (int128_t) correctly.

While developing the code, we found there is another problem/enhancement we need to do, so I created an issue for it https://github.com/facebookincubator/velox/issues/6248 This could be the next step.

### CODE REVIEW

**Comment by yiweiHeOSS** [comment]
It looks like my change to the 
```
[  FAILED  ] E2EFilterTest.longDecimalDictionary
[  FAILED  ] E2EFilterTest.longDecimalDirect
```

**Comment by yiweiHeOSS** [comment]
My change to the below code cuase failures in another testcase:
```
return HugeInt::build(Random::rand32(gen), Random::rand32(gen));
to        return HugeInt::build(Random::rand64(gen), Random::rand64(gen));
```
[  FAILED  ] E2EFilterTest.longDecimalDictionary
[  FAILED  ] E2EFilterTest.longDecimalDirect

So I modify the code to add a temp to see the velue in the debugger:
```
class HugeInt {
 public:
  static constexpr FOLLY_ALWAYS_INLINE int128_t
  build(uint64_t hi, uint64_t lo) {
    // GCC does not allow left shift negative value.
    int128_t temp = (static_cast<__uint128_t>(hi) << 64) | lo;
    return temp;
  }
```
lo: 10936052917074306677
hi: 15573954114018521923
temp: -52993621163942803648006051651297827211

The error message later in the test:
```
[ RUN      ] E2EFilterTest.longDecimalDictionary
/root/velox/velox/dwio/common/tests/E2EFilterTestBase.cpp:122: Failure
Value of: resultBatch->equalValueAt(batches[batchIndex].get(), i, rowIndex)
  Actual: false
Expected: true
Content mismatch at resultBatch 0 at index 0: expected: {-5299362116394280364800605165.1297827211, {-10962914699381264738035368224.7093712783}} actual: {-502284656663603218480.5461940619, {-524318847941508198011.4504906639}}
Google Test trace:
/root/velox/velox/dwio/common/tests/E2EFilterTestBase.cpp:322: No row group skip
/root/velox/velox/dwio/common/tests/E2EFilterTestBase.cpp:235: Failure
Value of: result->equalValueAt(expectedColumn, i, expectedRow)
  Actual: false
Expected: true
Content mismatch at 0 column 0: expected: -5299362116394280364800605165.1297827211 actual: -502284656663603218480.5461940619
Google Test trace:
/root/velox/velox/dwio/common/tests/E2EFilterTestBase.cpp:322: No row group skip
/root/velox/velox/dwio/common/tests/E2EFilterTestBase.cpp:235: Failure
Value of: result->equalValueAt(expectedColumn, i, expectedRow)
  Actual: false
Expected: true
```


So the expected value looks good, but the actual value is corrupted. Maybe there is a problem in the code of E2EFilterTest related to write/read.

**Comment by yiweiHeOSS** [comment]
expected：-52993621163942803648006051651297827211：
-100111110111100011001100111111011001011011000101001100101111000110100000111011010101011101000001001000111010010101100110001011
actual：-5022846566636032184805461940619：
-111111011001011011000101001100101111000110100000111011010101011101000001001000111010010101100110001011
The head bits are cut in actual 100111110111100011001100

**Comment by yiweiHeOSS** [comment]
Created an issue for the issue above https://github.com/facebookincubator/velox/issues/6317

File: velox/dwio/common/tests/utils/BatchMaker.cpp [hunk 1]
```diff
@@ -188,7 +188,7 @@ VectorPtr BatchMaker::createVector<TypeKind::HUGEINT>(
       size,
       gen,
       [&gen]() {
-        return HugeInt::build(Random::rand32(gen), Random::rand32(gen));
+        return HugeInt::build(Random::rand64(gen), Random::rand64(gen));
```
Comment (yingsu00) [inline velox/dwio/common/tests/utils/BatchMaker.cpp:191]: Add the commit of https://github.com/facebookincubator/velox/pull/6342 into this PR as a separate commit, and put it before adding the benchmarks

File: velox/dwio/common/tests/utils/FilterGenerator.h [hunk 2]
```diff
@@ -153,6 +153,10 @@ class ColumnStats : public AbstractColumnStats {
       case FilterKind::kIsNotNull:
         filter = std::make_unique<velox::common::IsNotNull>();
         break;
+      // Todo: add more filters here.
+      case FilterKind::kBytesRange:
```
Comment (yingsu00) [inline velox/dwio/common/tests/utils/FilterGenerator.h:157]: This should not be needed

File: velox/dwio/common/tests/utils/FilterGenerator.h [hunk 3]
```diff
@@ -334,6 +338,10 @@ class ColumnStats : public AbstractColumnStats {
         getIntegerValue(max), getIntegerValue(max), false);
   }
 
+  std::unique_ptr<Filter> makeBytesRangeFilter(const FilterSpec& filterSpec) {
```
Comment (yingsu00) [inline velox/dwio/common/tests/utils/FilterGenerator.h:341]: Remove

File: velox/dwio/common/tests/utils/FilterGenerator.h [hunk 4]
```diff
@@ -434,6 +446,14 @@ template <>
 std::unique_ptr<Filter> ColumnStats<double>::makeRangeFilter(
     const FilterSpec& filterSpec);
 
+template <>
+std::unique_ptr<Filter> ColumnStats<int128_t>::makeRangeFilter(
+    const FilterSpec& filterSpec);
+
+template <>
+std::unique_ptr<Filter> ColumnStats<StringView>::makeBytesRangeFilter(
```
Comment (yingsu00) [inline velox/dwio/common/tests/utils/FilterGenerator.h:454]: Remove

**Review by yingsu00 — COMMENTED** [review]
There shouldn't be any commits that just fix the format issues from previous commits. Please merge it with the previous one.

Add a separate commit to rewrite `ColumnStats<StringView>::makeRangeFilter` and add `ColumnStats<StringView>::makeRandomFilter`.

Separate the decimal benchmark and varchar benchmark into two commits.

Can you please add the benchmark result as a comment block at the end of the benchmark? Be sure to use release build, and include your hardware spec.

File: velox/dwio/common/tests/utils/FilterGenerator.cpp [hunk 5]
```diff
@@ -165,6 +165,40 @@ std::unique_ptr<Filter> ColumnStats<StringView>::makeRangeFilter(
     return std::make_unique<velox::common::IsNull>();
   }
 
+  int32_t lowerIndex;
```
Comment (yingsu00) [inline velox/dwio/common/tests/utils/FilterGenerator.cpp:181]: THe commit title is too long. It should be within 50 chars. Consider just use "rename ColumnStats<StringView>::makeRangeFilter"

Also, include in the commit message why this changes was made.

File: velox/dwio/parquet/tests/reader/ParquetReaderBenchmark.cpp [hunk 6]
```diff
@@ -399,675 +408,1503 @@ int main(int argc, char** argv) {
...
-run(List_Filter_100_Nulls_100_next_5k_plain)                49.41ms    20.24
-run(List_Filter_100_Nulls_100_next_10k_dict)                49.93ms    20.03
-run(List_Filter_100_Nulls_100_next_10k_Plain)               49.90ms    20.04
-run(List_Filter_100_Nulls_100_next_20k_dict)                49.38ms    20.25
-run(List_Filter_100_Nulls_100_next_20k_plain)               48.86ms    20.47
-run(List_Filter_100_Nulls_100_next_50k_dict)                48.91ms    20.45
-run(List_Filter_100_Nulls_100_next_50k_plain)               49.07ms    20.38
-run(List_Filter_100_Nulls_100_next_100k_dict)               49.25ms    20.31
-run(List_Filter_100_Nulls_100_next_100k_plain)              48.90ms    20.45
+run(ShortDecimalType_Filter_0_Nulls_0_next_5k_d            21.72ms     46.04
```
Comment (yingsu00) [inline velox/dwio/parquet/tests/reader/ParquetReaderBenchmark.cpp:418]: Actually, let's not remove the existing Intel results. Just add the the end of your AMD test results

File: velox/dwio/common/tests/utils/FilterGenerator.cpp [hunk 7]
```diff
@@ -165,6 +165,40 @@ std::unique_ptr<Filter> ColumnStats<StringView>::makeRangeFilter(
...
+      false,
+      false,
+      std::string(upper),
+      false,
+      false,
+      filterSpec.allowNulls_);
+}
+
+template <>
+std::unique_ptr<Filter> ColumnStats<StringView>::makeRandomFilter(
```
[the remark this answers is not in the document]
Comment (yiweiHeOSS) [reply velox/dwio/common/tests/utils/FilterGenerator.cpp:209]: It looks like it's hard to find all the places that are using the original "random filter". Is there a better way to do that? @yingsu00

[the remark this answers is not in the document]
Comment (yiweiHeOSS) [reply velox/dwio/common/tests/utils/FilterGenerator.cpp:209]: I made the change to solve this problem in a way.

File: velox/dwio/parquet/tests/reader/ParquetReaderBenchmark.cpp [hunk 8]
```diff
@@ -136,6 +152,11 @@ class ParquetReaderBenchmark {
     return rowReader;
   }
 
+  // This method is the place where we do the read opeartion.
```
Comment (yingsu00) [inline velox/dwio/parquet/tests/reader/ParquetReaderBenchmark.cpp:155]: opeartion -> operations

File: velox/dwio/parquet/tests/reader/ParquetReaderBenchmark.cpp [hunk 9]
```diff
@@ -136,6 +152,11 @@ class ParquetReaderBenchmark {
...
+  // scanSpec contains the setting of filter and null-filter.
```
Comment (yingsu00) [inline velox/dwio/parquet/tests/reader/ParquetReaderBenchmark.cpp:156]: What do you mean null-filter? The filter "IS NULL"? If that's the case, it's one kind of filters and you don't need to separate them.

File: velox/dwio/parquet/tests/reader/ParquetReaderBenchmark.cpp [hunk 10]
```diff
@@ -136,6 +152,11 @@ class ParquetReaderBenchmark {
...
+  // e.g. filter 30% filter 70% of data and 30% remain.
```
Comment (yingsu00) [inline velox/dwio/parquet/tests/reader/ParquetReaderBenchmark.cpp:157]: // e.g. filterRate = 30% (filterRateX100 = 30) means it would filter out 70% of rows and 30% remain.

File: velox/dwio/parquet/tests/reader/ParquetReaderBenchmark.cpp [hunk 11]
```diff
@@ -136,6 +152,11 @@ class ParquetReaderBenchmark {
...
+  // e.g. null-filter 70% filter 70% of data and 30% remain.
```
Comment (yingsu00) [inline velox/dwio/parquet/tests/reader/ParquetReaderBenchmark.cpp:158]: What do you mean null-filter? Why does it have opposite meaning with the filterRate? This is very confusing.

File: velox/dwio/parquet/tests/reader/ParquetReaderBenchmark.cpp [hunk 12]
```diff
@@ -208,6 +230,7 @@ class ParquetReaderBenchmark {
     // upperBound and lowerBound are introduced to estimate the result size.
     auto resultSize = read(rowType, scanSpec, nextSize);
 
+    // Calculate the expected number of rows after the filter and null-filter.
```
Comment (yingsu00) [inline velox/dwio/parquet/tests/reader/ParquetReaderBenchmark.cpp:233]: remove "and null-filter."

File: velox/dwio/common/tests/utils/FilterGenerator.h [hunk 13]
```diff
@@ -153,7 +153,14 @@ class ColumnStats : public AbstractColumnStats {
       case FilterKind::kIsNotNull:
         filter = std::make_unique<velox::common::IsNotNull>();
         break;
+      case FilterKind::kBytesRange:
+        filter = makeRangeFilter(filterSpec);
+        break;
       default:
+        if (type_->kind() == TypeKind::VARCHAR) {
```
Comment (yingsu00) [inline velox/dwio/common/tests/utils/FilterGenerator.h:160]: The following is slightly easier to read.
```
default:
        if (type_->kind() == TypeKind::VARCHAR) {
          filter = makeRandomFilter(filterSpec);
        } else {
          filter = makeRangeFilter(filterSpec);
        }
        break;
```

**Review by yingsu00 — COMMENTED** [review]
Please merge the last commit "Change the callers of original makeRangeFilter()." into the first one. 
Many of the commit message lines are too long. Please take a look at https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53 and update them accordingly.

File: velox/dwio/parquet/tests/reader/ParquetReaderBenchmark.cpp (hunk 11 above)
Comment (yiweiHeOSS) [reply velox/dwio/parquet/tests/reader/ParquetReaderBenchmark.cpp:158]: I just want to explain what filterRateX100 and nullsRateX100 are. It is hard to a new person to understand these rates. The nullsRateX100 has opposite meaning with the filterRateX100.

**Comment by yiweiHeOSS** [comment]
Simplified/shortened the commit messages and made the change according to the comments.

**Review by yingsu00 — COMMENTED** [review]
Can you merge "Merge branch 'main' into addTypes" into the proper commit of the 4 previous commits? It's not good to have a commit that just merge branch.

**Review by yingsu00 — APPROVED** [review]

**Comment by czentgr** [comment]
@yiweiHeOSS please rebase to the lastest. Looks like this is over 2 weeks old already. Any reason the 4 commits were not squashed? Best to squash and provide a comprehensive description for the final commit (as found in this PR description). Please let me know if you need help with that.
@Yuhta Would you please be able to review this PR? Thank you!

**Comment by yiweiHeOSS** [comment]
Hi @Yuhta , just a reminder, the PR https://github.com/facebookincubator/velox/pull/6342 also needs to merge. It is a dependency of this PR, not sure if it is merged already.
